### PR TITLE
ci: added miri job

### LINKED ISSUE #1088: ci: Integrate miri job
### Describe the feature

add a [miri](https://github.com/rust-lang/miri) ci job to check for undefined behaviour

See also:

https://github.com/rust-lang/miri#running-miri-on-ci

or tokio's miri job:

https://github.com/tokio-rs/tokio/blob/88b1eb54fb66461b9f3524f4b5316241a019279a/.github/workflows/ci.yml#L193-L211

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xqft**
Closes #1088

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -222,3 +222,35 @@ jobs:
...
+
+  miri:
+    name: miri
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v3
+
+      - name: Install toolchain
+        uses: dtolnay/rust-toolchain@nightly
```
Comment (mattsse) [inline .github/workflows/ci.yml:234]: I believe miri can be added directly:

```
        with:
          components: miri
```

**Review by mattsse — CHANGES_REQUESTED** [review]
looks like it has problems with the fuzz tests, doesn't look like UB.
so perhaps we annotate them with miri ignore?

maybe we don't want to run this on every PR
wdyt @onbjerg ?

**Comment by xqft** [comment]
Yep, there's problems with `#[test_fuzz]` and `#[tokio::test]` as specified by tokio's miri job. I've been debugging this for a while and the solution doesn't seem so trivial, I'm not sure how tokio solved this, they just seem to accepted the fact that some tests may fail by using `--no-fail-fast`.

**Comment by mattsse** [comment]
I believe the tokio crate has some comments regarding miri and tokio::test tokio::main in its codebase

**Comment by xqft** [comment]
I did some investigation with the help of @fkrause98 and we don't see a way out from this right now:

- The general problem is that miri doesn't support many functions and syscalls which are implemented in libraries like `tokio` and `test-fuzz`. In particular macros like `#[test_fuzz]` and `#[tokio::test]` derive unsupported functionality by miri. This means that any test executed with these will fail.
- We think tokio's way of dealing with this is by only testing lib tests by using the `--lib` flag, not the bin crate one's which is where they implement the most unsupported functionality. There's only one test in their lib which would fail and they make miri ignore it by the use of `#[cfg_attr(miri, ignore)]`. Other two tests use this macro too but it's because they run too slow on miri. They also use the `--no-fail-fast` flag I think because of convenience, so only one failed miri test doesn't stop the whole job. I don't think it's related to the problem.

We could theoretically implement `#[cfg_attr(miri, ignore)]` in every test that fails because of unsupported, but:
- They're too many. Every async tokio test will need to be ignored, also every fuzz test, and I found some of these which are generated by other functions, something that adds complexity.
- Everytime a new test is implemented, we need to be careful to not use unsupported functionality or to implement the ignore macro. This creates burden for people who didn't jump into this rabbit hole yet.

Possible ways of action:
- Using the `-Zmiri-panic-on-unsupported` flag and find a way to filter out every test which fails because of `unsupported functionality` ([this is relevant](https://github.com/rust-lang/miri/issues/1908#issuecomment-961431629))
- Could it be possible to make miri ignore every test and enable it only on those we think could have UB?

New ideas and insight about this are welcome, we may be missing something

**Comment by onbjerg** [comment]
I think in extension to above, we should probably not fail CI if this job fails seeing as miri is unstable/experimental? @mattsse

**Comment by xqft** [comment]
@gakonst using the `-Zmiri-panic-on-unsupported` flag every test that fails because of a panic should be about unsupported functionality, and every other which fails with an error should be because of UB. Also we can filter just by the `unsupported operation` string.

**Comment by xqft** [comment]
I arrived to this solution by creating a custom github action, it works by:

- Outputting the tests results in json, parsing and formatting these in a readable way with [jq](https://stedolan.github.io/jq/).
- Tests status can be `ok`, `failed` or `unsupported`. Unsupported tests' errors are supressed by the use of `-Zmiri-panic-on-unsupported`, although some tests seem to bypass this and throw errors on stdout anyway instead of just panicking.
- Because of the unsupported tests, the job would always fail, so I piped the output to a script that exits with status `101` only if a regex matches that a test has status `failed` (because of UB). Else it exits with 0.
- Proptests seem to hang with miri and get sigkilled

This is all really crumbly, the action facilitates things a bit but miri is slow and unstable (oficially experimental), I would only recommend to run this job maybe in a set apart branch that's up to date with main. Maybe a CI job with miri should be scraped entirely and just run it locally with a limited amount of tests each time.

**Comment by gakonst** [comment]
oof it looks like it doesn't like FFI either

**Comment by onbjerg** [comment]
We can integrate this with nextest which is pretty nice: https://nexte.st/book/miri.html

**Comment by xqft** [comment]
I discarded nextest because it [doesn't yet support tests run in json output](https://nexte.st/book/machine-readable.html#running-tests)

**Comment by gakonst** [comment]
@xqft what's the plan here? I think if we don't have a "clean" way of integrating with success, let's close and revisit later.

**Comment by gakonst** [comment]
agreed - thank you for taking the time in digging into this.

**Pull request closed without merging**